arduino/cores/packagemanager/package_manager.go

```diff
@@ -50,20 +50,62 @@ type PackageManager struct {
 	userAgent              string
 }
 
+// Builder is used to create a new PackageManager. The builder
+// has methods to load patforms and tools to actually build the PackageManager.
+// Once the PackageManager is built, it cannot be changed anymore.
+type Builder struct {
+	*PackageManager
+}
+
```

Comment:
What about using an interface instead of using this encapsulation?

Reply:
The solution is to not use either interface or encapsulation... https://github.com/arduino/arduino-cli/pull/1828/commits/721e4077cface533dcf3ee69a929e9945eae5cfd

Builder, Explorer, and PackageManager are 3 different entities that do not share any common interface. Encapsulation turned out useful during development to keep the number of syntax errors under control, but once finished it makes no more sense to keep this hierarchy, instead it's even dangerous because trivial errors may pass unnoticed as demonstrated by the commit above.